test/elixir_analyzer/exercise_test/common_checks/function_annotation_order_test.exs

```diff
@@ -0,0 +1,573 @@
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          def x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == [
+             {:fail,
+              %Comment{
+                type: :informative,
+                name: Constants.solution_function_annotation_order(),
+                comment: Constants.solution_function_annotation_order(),
+                params: %{
+                  expected: """
+                  @doc
+                  @spec x
+                  def x
+                  """,
+                  actual: """
+                  @spec x
+                  @doc
+                  def x
+                  """
+                }
+              }}
+           ]
+  end
+
+  test "works for def, defp, defmacro, defmacrop, defguard, and defguardp" do
+    order_error = fn op, name ->
+      [
+        {:fail,
+         %Comment{
+           type: :informative,
+           name: Constants.solution_function_annotation_order(),
+           comment: Constants.solution_function_annotation_order(),
+           params: %{
+             expected: """
+             @doc
+             @spec #{name}
+             #{op} #{name}
+             """,
+             actual: """
+             @spec #{name}
+             @doc
+             #{op} #{name}
+             """
+           }
+         }}
+      ]
+    end
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          def x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("def", "x")
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          defp x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("defp", "x")
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          defmacro x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("defmacro", "x")
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          defmacrop x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("defmacrop", "x")
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          defguard x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("defguard", "x")
+
+    ast =
+      quote do
+        defmodule Test do
+          @spec x()
+          @doc ""
+          defguardp x()
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == order_error.("defguardp", "x")
+  end
+
+  test "non related definitions will not make crash" do
+    ast =
+      quote do
+        defmodule Test do
+          @doc ""
+          def x
+
+          @spec y
+          def y
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == []
+  end
```

Comment:
Can we also test this? I didn't see it in the suite, but if it already exists, feel free to disregard:

```elixir
  test "another non related definition will not make crash" do
    ast =
      quote do
        defmodule Test do
          @spec x
          def x

          @doc ""
          def y
        end
      end

    assert FunctionAnnotationOrder.run(ast) == []
  end
```